src/violation.rs

```diff
@@ -1,15 +1,98 @@
-// TODO: Implement Violations for constraints
-#[derive(Debug, Clone)]
-pub struct Violations {
-    violations: Vec<Violation>,
+/// Represents Violations found during validation of values based on schema types
+pub trait Violations {
+    /// adds a violation into the collection of [Violation]s
+    fn add_violation(&mut self, violation: ViolationImpl);
+
+    /// Returns `true` if no violations were found, otherwise `false`
+    fn is_valid(&self) -> bool;
+
+    /// Returns the [Violation]s that were found during validation
+    fn violations(&self) -> &[ViolationImpl];
+
+    /// Creates a checkpoint to track the number of violations from this point onwards
+    fn checkpoint(&self) -> Checkpoint;
```

Comment:
question—If the `validate` functions returned a `Option<Violation>` rather than using side-effects, we wouldn't need the `checkpoint` function, and it _seems_ like the net result would be a simpler implementation. Do you know if it's possible to do this, and if it would actually result in a simpler or more complicated implementation?

Reply:
I feel it might complicate this if I use an `Option<Violation>` here because at the end we need an accumulated result of all `Violation`s and if every time a new `Violation` is created then it needs to be added to the `Violations` object after its returned. Also, for container type validations the children violations needs to be added as well.

Reply:
@popematt Do we know why `ion-schema-kotlin` tries to report all possible violations rather than just the first one?

Reply:
Basically, we don't think there's ever a use case for knowing only one of the violations. Uses cases fall into two categories:
1. You only want to find out whether the data is valid or not. If this is your use case, then you want _none_ of the violations—just a boolean answer. In `ion-schema-kotlin` you can call `isValid()` for this.
2. You want to explain why the data is invalid. If this is your use case, then you should want _all_ of the violations at once. Firstly, because it's really annoying to work with an API that will only expose 1 error at a time as you're trying to fix things. Secondly, because sometimes you need more than one (child) violation to explain what is wrong. (Eg. the `one_of` constraint when it matches two types.)

That being said, there is really only ever one top-level violation ("value does not match type ___"), and the individual reasons why it does not match are children of that top level violation. However, `ion-schema-kotlin` models violations in (what I believe to be) a confusing way. 
* The object returned by `validate()` is a `Violations`.
* `Violations` has no message of its own and is an aggregation of `Violation`. 
* `Violation` extends `Violations`, but has a message, and aggregates `Violation` and `ViolationChild`. It represents a specific reason why the data is invalid.
* `ViolationChild` extends `Violations`, but has a message and optionally a field name or list index, and aggregates `Violation` and `ViolationChild`. It represents a specific location in the data that is invalid.

So, `ion-schema-kotlin` attempts to model violations as a forest when it should probably be a tree.